File: festival-kotlin-web/src/main/java/cn/bdqfork/web/route/message/resolver/AbstractParameterResolver.java
package cn.bdqfork.web.route.message.resolver;

import io.vertx.core.MultiMap;
import io.vertx.core.http.HttpMethod;
import io.vertx.core.http.HttpServerRequest;
import io.vertx.ext.web.RoutingContext;

import java.lang.reflect.Parameter;
import java.util.Map;

/**
 * @author bdq
 * @since 2020/2/11
 */
public abstract class AbstractParameterResolver implements ParameterResolver {

    private AbstractParameterResolver next;

    public void setNext(AbstractParameterResolver next) {
        this.next = next;
    }

    @Override
    public Object resolve(Parameter parameter, RoutingContext routingContext) {
        if (resolvable(parameter)) {
            return doResolve(parameter, routingContext);
        }
        if (next != null) {
            return next.resolve(parameter, routingContext);
        }
        return null;
    }

    protected abstract Object doResolve(Parameter parameter, RoutingContext routingContext);

    protected abstract boolean resolvable(Parameter parameter);

    protected MultiMap resolveParams(RoutingContext routingContext) {
        Map<String, String> pathParams = routingContext.pathParams();
        if (routingContext.request().method() == HttpMethod.GET) {
            return routingContext.queryParams().addAll(pathParams);
        } else {
            HttpServerRequest httpServerRequest = routingContext.request();
            return httpServerRequest.formAttributes().addAll(pathParams);
        }
    }
}

File: festival-kotlin-web/src/main/kotlin/cn/bdqfork/kotlin/web/route/message/resolver/AbstractParameterResolver.kt
package cn.bdqfork.kotlin.web.route.message.resolver

import io.vertx.core.MultiMap
import io.vertx.core.http.HttpMethod
import io.vertx.ext.web.RoutingContext
import java.lang.reflect.Parameter

/**
 * @author bdq
 * @since 2020/2/11
 */
abstract class AbstractParameterResolver : ParameterResolver {
    private var next: AbstractParameterResolver? = null

    fun setNext(next: AbstractParameterResolver?) {
        this.next = next
    }

    @Throws(Exception::class)
    override fun resolve(parameter: Parameter, routingContext: RoutingContext): Any? {
        if (resolvable(parameter,routingContext)) {
            return doResolve(parameter, routingContext)
        }
        return if (next != null) {
            next!!.resolve(parameter, routingContext)
        } else null
    }

    @Throws(Exception::class)
    protected abstract fun doResolve(parameter: Parameter, routingContext: RoutingContext): Any?
    protected abstract fun resolvable(parameter: Parameter,routingContext: RoutingContext): Boolean

    protected fun resolveParams(routingContext: RoutingContext): MultiMap {
        val pathParams = routingContext.pathParams()
        return if (routingContext.request().method() == HttpMethod.GET) {
            routingContext.queryParams().addAll(pathParams)
        } else {
            val httpServerRequest = routingContext.request()
            httpServerRequest.formAttributes().addAll(pathParams)
        }
    }
}